lib/heex_formatter/formatter.ex

```diff
@@ -17,42 +29,30 @@ defmodule HeexFormatter.Phases.Format do
   The following string will be returned:
 
   "<section>\n  <div>\n    <h1>\n      Hello\n    </h1>\n  </div>\n</section>\n"
-
-  Notice that this string is formatted. So this is supposed to be the last
-  step before writing it to a file.
   """
-
-  # Use 2 spaces for a tab
-  @tab "  "
-
-  @preformatted_tags ~w(script code style pre)
-
-  # Line length of opening tags before splitting attributes onto their own line
-  @default_line_length 98
-
-  @spec run(list(), Keyword.t()) :: String.t()
-  def run(tokens, opts) do
+  def format(tokens, opts) do
     initial_state = %{
-      html: "",
+      buffer: [],
       previous_token: nil,
       indentation: 0,
       line_length: opts[:heex_line_length] || opts[:line_length] || @default_line_length,
-      formatter_opts: opts
+      formatter_opts: opts,
+      mode: :normal
     }
 
-    result =
-      Enum.reduce(tokens, initial_state, fn token, state ->
-        new_state = token_to_string(token, state)
-
-        # Set the previous token so we can check it to know how the current tag should
-        # be formatted.
-        %{new_state | previous_token: token}
-      end)
+    tokens
+    |> Enum.reduce(initial_state, fn token, state ->
+      new_state = token_to_string(token, state)
+      %{new_state | previous_token: token}
+    end)
+    |> buffer_to_string()
+  end
 
-    # That is because we are adding "\n" every time a tag is open. Then we need to extract
-    # "\n" from the first line and put this in the end of the line.
-    "\n" <> html = result.html
-    html <> "\n"
+  defp buffer_to_string(state) do
```

Comment:
Change the logic to not use `<>` but append items with `[left, right]` and apply `IO.io_data_to_binary` in the final result.

TIL:  We don't want to use `<>` within a loop to append binary 😅